### PR TITLE
Add Awesome/Monad style automatic layouts to Sway

### PR DESCRIPTION
**Pull request by willakat**
Hi all,

I've been wanting to try out a Wayland setup and as an Awesome (and ex-i3) user, I gave Sway a spin. I quickly missed the (in my opinion) easy way with which Awesome just sorts out all the nitty-gritty of laying out your windows for you with a few simple pre-defined layouts that usually cover my limited needs. Anyway, long story short, I added the feature to Sway and find it quite pleasant to use. I took the opportunity to refactor some of the layout code. I'll probably add an includable config file to provide a setup that roughly matches Awesome's.

Enjoy!

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -10,6 +11,10 @@ irc.freenode.net).
 
 [More screenshots](https://github.com/SirCmpwn/sway/wiki/Screenshots-of-Sway)
 
+This fork attempts to add
```
Comment (Hummer12007) [inline README.md]: >This fork

Comment (SirCmpwn) [inline README.md]: Yeah just revert all of the readme changes

File: include/sway/layout.h [hunk 2]
```diff
@@ -75,4 +75,7 @@ void swayc_log(log_importance_t verbosity, swayc_t *cont, const char* format, ..
  */
 enum swayc_layouts default_layout(swayc_t *output);
 
+inline bool is_auto_layout(enum swayc_layouts layout) {
```
Comment (SirCmpwn) [inline include/sway/layout.h]: Needs to move out of the header and into a file (no code in headers, just don't make it inline)

File: sway/commands/layout.c [hunk 3]
```diff
@@ -49,11 +49,86 @@ struct cmd_results *cmd_layout(int argc, char **argv) {
...
+		} else if (strcasecmp(argv[0], "auto_top") == 0) {
+			swayc_change_layout(parent, L_AUTO_TOP);
+		} else if (strcasecmp(argv[0], "auto_bot") == 0) {
+			swayc_change_layout(parent, L_AUTO_BOTTOM);
+		} else if (strcasecmp(argv[0], "incnmaster") == 0)  {
+			if ((error = checkarg(argc, "layout incnmaster",
+					      EXPECTED_EQUAL_TO, 2))) {
+				return error;
+			}
+			int inc = (int) strtol(argv[1], NULL, 10);
```
Comment (SirCmpwn) [inline sway/commands/layout.c]: Can you add error handling for invalid ints here?

```c
char *end;
int inc = (int)strtol(argv[1], &end, 10);
if (*end) {
    // error
}
```

Also note no space between (int) and strtol

File: sway/commands/layout.c [hunk 4]
```diff
@@ -49,11 +49,86 @@ struct cmd_results *cmd_layout(int argc, char **argv) {
...
+		} else if (strcasecmp(argv[0], "auto") == 0) {
+			if ((error = checkarg(argc, "auto", EXPECTED_EQUAL_TO, 2))) {
+				return error;
+			}
+			swayc_t *container = get_focused_view(swayc_active_workspace());
+			swayc_t *parent = container->parent;
+			enum swayc_layouts layout;
+			if (strcasecmp(argv[1], "next") == 0) {
+				if (is_auto_layout(parent->layout) && parent->layout < L_AUTO_LAST) {
+					layout = parent->layout + 1;
```
Comment (SirCmpwn) [inline sway/commands/layout.c]: Can you add another function somewhere that uses a switch case or something, given a layout, to find the next auto layout? I'd rather not rely on the order of and value of the enum values.

File: sway/commands/layout.c [hunk 5]
```diff
@@ -49,11 +49,86 @@ struct cmd_results *cmd_layout(int argc, char **argv) {
...
+					layout = parent->layout - 1;
+				} else {
+					layout = L_AUTO_LAST;
+				}
+			} else {
+				return cmd_results_new(CMD_FAILURE, "layout auto",
+						       "Must be one of <prev|next>.");
+			}
+			swayc_change_layout(parent, layout);
+		} else if (strcasecmp(argv[0], "promote") == 0) {
```
Comment (SirCmpwn) [inline sway/commands/layout.c]: Would prefer this as `move promote`

File: sway/commands/resize.c [hunk 6]
```diff
@@ -62,225 +62,230 @@ static bool resize_floating(int amount, bool use_width) {
...
+
+}
+
+/**
+ * given the index of a container's child, return the index of the first child of the group
+ * which index is a member of.
+ */
+static int group_start_index(swayc_t *container, int index) {
+	if ((index < 0) || (! is_auto_layout(container->layout)) ||
+	    ((uint_fast32_t) index < container->nb_master)) {
```
Comment (SirCmpwn) [inline sway/commands/resize.c]: Weird indent here, also weird extra parenthesis

```c
if (index < 0 || !is_auto_layout(container->layout)
	|| (ssize_t)index < container->nb_master)) {
}
```

File: sway/commands/resize.c [hunk 7]
```diff
@@ -62,225 +62,230 @@ static bool resize_floating(int amount, bool use_width) {
...
+		return 0;
+	} else {
+		uint_fast32_t grp_sz = slave_count(container) / container->nb_slave_groups;
+		uint_fast32_t remainder = slave_count(container) % container->nb_slave_groups;
```
Comment (SirCmpwn) [inline sway/commands/resize.c]: Use size_t instead of uint_fast32_t

File: sway/commands/resize.c [hunk 8]
```diff
@@ -62,225 +62,230 @@ static bool resize_floating(int amount, bool use_width) {
...
+ * given the index of a container's child, return the index of the first child of the group
+ * that follows the one which index is a member of.
+ * This makes the function usable to walk through the groups in a container.
+ */
+static int group_end_index(swayc_t *container, int index) {
+	if (index < 0 || ! is_auto_layout(container->layout)) {
+		return container->children->length;
+	} else {
+		int nxt_idx;
+		if ((uint_fast32_t)index < container->nb_master) {
```
Comment (SirCmpwn) [inline sway/commands/resize.c]: That goes for everywhere, s/uint_fast32_t/size_t/g

File: sway/commands/resize.c [hunk 9]
```diff
@@ -62,225 +62,230 @@ static bool resize_floating(int amount, bool use_width) {
...
+ */
+static inline uint_fast32_t group_count(swayc_t *container) {
+	return MIN(container->nb_slave_groups, slave_count(container)) + (container->nb_master ? 1 : 0);
+}
+
+/**
+ * return the index of the Group containing <index>th child of <container>.
+ * The index is the order of the group along the container's major axis (starting at 0).
+ */
+static uint_fast32_t group_index(swayc_t *container, int index) {
```
Comment (SirCmpwn) [inline sway/commands/resize.c]: Can you prefix all of these auto related functions with auto_?

File: sway/commands/resize.c [hunk 10]
```diff
@@ -62,225 +62,230 @@ static bool resize_floating(int amount, bool use_width) {
...
+	uint_fast32_t nb_before = 0;
+	uint_fast32_t nb_after = 0;
+
+	// 1. Identify a container ancestor that will allow the focused child to grow in the requested
+	//    direction.
+	while (container->parent) {
+		parent = container->parent;
+		if ((parent->children && parent->children->length > 1) &&
+		    (is_auto_layout(parent->layout) || (use_width ? parent->layout == L_HORIZ :
+							parent->layout == L_VERT))) {
```
Comment (SirCmpwn) [inline sway/commands/resize.c]: More weird indentation, use tabs

File: sway/commands/resize.c [hunk 11]
```diff
@@ -62,225 +62,230 @@ static bool resize_floating(int amount, bool use_width) {
...
+							 use_width ? WLC_RESIZE_EDGE_RIGHT : WLC_RESIZE_EDGE_BOTTOM);
 				}
 			}
-			arrange_windows(swayc_active_workspace(), -1, -1);
+			i = next_i;
 		}
-		return true;
+		// Recursive resize does not handle positions, let arrange_windows
+		// take care of that.
+		arrange_windows(swayc_active_workspace(), -1, -1);
```
Comment (SirCmpwn) [inline sway/commands/resize.c]: This whole function has now become pretty complicated. Can you split it up into several?

File: sway/container.c [hunk 12]
```diff
@@ -960,11 +962,28 @@ swayc_t *swayc_tabbed_stacked_parent(swayc_t *con) {
 }
 
 swayc_t *swayc_change_layout(swayc_t *container, enum swayc_layouts layout) {
+	// if layout change modifies the auto layout's major axis, swap width and height
+	// to preserve current ratios.
+	if (is_auto_layout(layout) && is_auto_layout(container->layout)) {
+		enum swayc_layouts prev_major = (container->layout == L_AUTO_LEFT ||
```
Comment (SirCmpwn) [inline sway/container.c]: Unnecessary parens

File: sway/layout.c [hunk 13]
```diff
@@ -256,7 +260,7 @@ void move_container(swayc_t *container, enum movement_direction dir) {
 		layout = L_VERT;
 	} else if (dir == MOVE_LEFT || dir == MOVE_RIGHT) {
 		layout = L_HORIZ;
-	} else {
+	} else if (! (dir == MOVE_NEXT || dir == MOVE_PREV)) {
```
Comment (SirCmpwn) [inline sway/layout.c]: No space after !

File: sway/layout.c [hunk 14]
```diff
@@ -279,19 +283,53 @@ void move_container(swayc_t *container, enum movement_direction dir) {
 		sway_log(L_DEBUG, "container:%p, parent:%p, child %p,",
 				container,parent,child);
 		if (parent->layout == layout
+			|| layout == L_NONE /* accept any layout for next/prev direction */
```
Comment (SirCmpwn) [inline sway/layout.c]: This won't accept any layout, right? It'll accept no layout, which isn't a thing that happens in practice (things like views have no layout)

File: sway/layout.c [hunk 15]
```diff
@@ -769,21 +808,42 @@ void update_geometry(swayc_t *container) {
...
+			      const double height, const int start,
+			      const int end);
+static void apply_tabbed_or_stacked_layout(swayc_t *container, double x,
+					   double y, double width,
+					   double height);
+
+static void apply_auto_layout(swayc_t *container, const double x, const double y,
+			      const double width, const double height,
+			      enum swayc_layouts group_layout,
+			      bool master_first);
```
Comment (SirCmpwn) [inline sway/layout.c]: Please make the indentation on these consistent

File: sway/layout.c [hunk 16]
```diff
@@ -895,136 +955,322 @@ static void arrange_windows_r(swayc_t *container, double width, double height) {
...
+
+	// size of the next group along the major axis.
+	double group_dim;
+
+	// height and width of next group to be laid out.
+	const double *group_h, *group_w;
+
+	switch(group_layout) {
+	default:
+		sway_log(L_ERROR, "Unknown layout type (%d) used in %s()",
```
Comment (SirCmpwn) [inline sway/layout.c]: Make this L_DEBUG. It's programmer error, right? Not user error? And is a recoverable condition?

File: sway/layout.c [hunk 17]
```diff
@@ -895,136 +955,322 @@ static void arrange_windows_r(swayc_t *container, double width, double height) {
...
+	// pos of the next group to layout along the major axis
+	double pos;
+
+	// size of the next group along the major axis.
+	double group_dim;
+
+	// height and width of next group to be laid out.
+	const double *group_h, *group_w;
+
+	switch(group_layout) {
```
Comment (SirCmpwn) [inline sway/layout.c]: Space after switch

File: sway/layout.c [hunk 18]
```diff
@@ -895,136 +955,322 @@ static void arrange_windows_r(swayc_t *container, double width, double height) {
...
+			break;
+		case L_HORIZ:
+			apply_horiz_layout(container, *group_x, *group_y, *group_w, *group_h, start, end);
+			break;
 		}
+
+		/* update position for next group */
+		pos += group_dim;
 	}
 }
```
Comment (SirCmpwn) [inline sway/layout.c]: Could this function be split up a bit?

**Comment by SirCmpwn** [comment]
Overall these changes look good. I like the refactored layout code. You should do a pass on this with the style guide in hand.

There's a lot of complexity I would like to see simplified if you can spare the time to go through everything and see if you can find ways to simplify it (optional).

These are some huge changes, so I'm not going to merge them right away. I'm going to start running them locally for a few days and see if they break any existing functionality. Would appreciate it if any Awesome users could try this and verify that it behaves appropriately as well.

Could you make a wiki page explaining how to configure sway to behave like Awesome?

File: sway/commands/layout.c (hunk 4 above)
Comment (SirCmpwn) [inline sway/commands/layout.c]: Nevermind, let's not

**Comment by SirCmpwn** [comment]
Pushed a couple of fixup commits to your branch

**Comment by SirCmpwn** [comment]
Resize commands behave strangely in auto modes (for obvious reasons, I guess, but would be nice if they did work) - resize commands work for slave views, but not masters. Mouse resize does not work in auto layouts.

When leaving an auto layout for an i3 layout, should we reset the incncols and such?

**Comment by SirCmpwn** [comment]
Instead of `layout incnmaster [n]` how about `layout auto [master|slave] [add|set] [n]`

**Comment by SirCmpwn** [comment]
How about making `layout auto_left` into `layout auto left`, then we can fold all auto-related layout commands into one sub-command and move it into a seperate function?

**Comment by SirCmpwn** [comment]
Clarification - resizing with resize command does work fine in auto layouts, it was my error. Resizing with mouse does not.

**Comment by SirCmpwn** [comment]
Also, have you pitched this feature to i3?

File: README.md (hunk 1 above)
Comment (willakat) [inline README.md]: done

File: sway/commands/layout.c (hunk 3 above)
Comment (willakat) [inline sway/commands/layout.c]: done

File: sway/commands/layout.c (hunk 5 above)
Comment (willakat) [inline sway/commands/layout.c]: renamed to "move first" (I actually had this change pending ... so I completely agree!)

File: sway/commands/resize.c (hunk 6 above)
Comment (willakat) [inline sway/commands/resize.c]: trying out "spacemacs", haven't yet looked into having it format using tabs only. For now fixed it up manually.

File: sway/commands/resize.c (hunk 7 above)
Comment (willakat) [inline sway/commands/resize.c]: done

File: sway/commands/resize.c (hunk 8 above)
Comment (willakat) [inline sway/commands/resize.c]: done

File: sway/commands/resize.c (hunk 9 above)
Comment (willakat) [inline sway/commands/resize.c]: done

File: sway/commands/resize.c (hunk 10 above)
Comment (willakat) [inline sway/commands/resize.c]: fixed

File: sway/commands/resize.c (hunk 11 above)
Comment (willakat) [inline sway/commands/resize.c]: on todo list...

File: sway/layout.c (hunk 15 above)
Comment (willakat) [inline sway/layout.c]: attempted to make things more consistent (used tabs exclusively)

File: sway/layout.c (hunk 16 above)
Comment (willakat) [inline sway/layout.c]: done

File: sway/layout.c (hunk 17 above)
Comment (willakat) [inline sway/layout.c]: fixed

File: sway/layout.c (hunk 18 above)
Comment (willakat) [inline sway/layout.c]: definitely. on todo list.

**Comment by willakat** [comment]
- wrt the "move" commands, I had something that seemed sensible at first but since I don't really use "layouts within layouts" other than for testing I'm unsure the behavior is intuitive (I also need to check whether I accidently broke it at some point).
- incnmaster/incncol are just a reuse of the same commands in awesome. So it's simple enough to switch to a different syntax. However incncol isn't strictly equivalent to changing the number of slaves (the number of slaves is simply #children - #masters), ncol is the number of "groups" over which all the slaves are distributed.
- auto_left to "auto left" does look better from a user standpoint. I'll add that to my todo list.
- resizing using the mouse isn't implemented. I was planning on having a look at it, but I first need a working understanding of how its handled in normal Sway.
- I haven't pitched this to i3 and actually hadn't really considered it (a friend of mine once was hoping to see some changes pushed into i3 to help with better KDE integration and these were rejected because they weren't in line with intended direction of development. Kind of a turn-off)

**Comment by SirCmpwn** [comment]
>wrt the "move" commands, I had something that seemed sensible at first but since I don't really use "layouts within layouts" other than for testing I'm unsure the behavior is intuitive (I also need to check whether I accidently broke it at some point).

Even for moving with the auto layout, that is. Nested layouts is something that needs to work (and so far as I can tell your changes don't introduce issues with that).

>incnmaster/incncol are just a reuse of the same commands in awesome. So it's simple enough to switch to a different syntax. However incncol isn't strictly equivalent to changing the number of slaves (the number of slaves is simply #children - #masters), ncol is the number of "groups" over which all the slaves are distributed.

We aren't trying to be compatible with Awesome configuration here (that'd be a fool's errand), but I'm okay with offering similar features like this PR does. We should stick to idiomatically i3 style commands.

>I haven't pitched this to i3 and actually hadn't really considered it (a friend of mine once was hoping to see some changes pushed into i3 to help with better KDE integration and these were rejected because they weren't in line with intended direction of development. Kind of a turn-off)

Well, you'll never know if you don't try.

**Comment by SirCmpwn** [comment]
Running swaygrab causes this branch to crash

```
Program terminated with signal SIGSEGV, Segmentation fault.
#0  0x00000000004276b0 in ipc_json_describe_view (c=0x2d45a40, object=0x2d55170) at /home/sircmpwn/sources/sway/sway/ipc-json.c:158
158		const char *layout = (c->parent->type == C_CONTAINER) ?
(gdb) bt
#0  0x00000000004276b0 in ipc_json_describe_view (c=0x2d45a40, object=0x2d55170) at /home/sircmpwn/sources/sway/sway/ipc-json.c:158
#1  0x0000000000427d3d in ipc_json_describe_container (c=0x2d45a40) at /home/sircmpwn/sources/sway/sway/ipc-json.c:228
#2  0x000000000042893a in ipc_json_describe_container_recursive (c=0x2d45a40) at /home/sircmpwn/sources/sway/sway/ipc-json.c:393
#3  0x0000000000428ada in ipc_json_describe_container_recursive (c=0x649140 <root_container>) at /home/sircmpwn/sources/sway/sway/ipc-json.c:416
#4  0x0000000000429d45 in ipc_client_handle_command (client=0x2545b50) at /home/sircmpwn/sources/sway/sway/ipc-server.c:461
#5  0x0000000000429301 in ipc_client_handle_readable (client_fd=83, mask=1, data=0x2545b50) at /home/sircmpwn/sources/sway/sway/ipc-server.c:245
#6  0x00007fa4ea98cd32 in wl_event_loop_dispatch () at /usr/lib/libwayland-server.so.0
#7  0x00007fa4ea98b1da in wl_display_run () at /usr/lib/libwayland-server.so.0
#8  0x00007fa4eb27f6d7 in wlc_run () at /home/sircmpwn/sources/wlc/src/wlc.c:260
#9  0x0000000000430809 in main (argc=2, argv=0x7ffcb333adf8) at /home/sircmpwn/sources/sway/sway/main.c:347
```

core dump available upon request

**Comment by SirCmpwn** [comment]
Been running this for a while and haven't encountered any issues aside from the ones already mentioned. This should be good to merge once the outstanding concerns are resolved.

File: sway/container.c (hunk 12 above)
Comment (willakat) [inline sway/container.c]: done

File: sway/layout.c (hunk 14 above)
Comment (willakat) [inline sway/layout.c]: "layout == L_NONE" was to broad, replaced by "layout == L_NONE && parent->type == C_CONTAINER". This is intended to accept the MOVE_PREV/MOVE_NEXT directions regardless of the actual container layout.

File: sway/layout.c (hunk 15 above)
Comment (willakat) [inline sway/layout.c]: grepped through the code for mixed tab/space indents and removed the ones I had put in.

**Comment by willakat** [comment]
I've been playing around with swaygrab but haven't encountered any SEGVs. I'm stumped as to what caused the parent pointer to get mangled. I could poke around the core file (if you kept it) and it would be interesting if this is a/ reproducible and b/ what sequence of actions preceded the crash.

As for the rest of the comments on the patch, I think I'm only missing the simplification bit (a few have been made, but nothing significant).

Regarding reseting nb_master/nb_slave_groups when switching layouts, I would argue against since it's very convenient to switch back and forth to, say a tabbed layout, to use all the available space, and recover the child layout as it was prior to the switch.

I also still need to push a sample Awesome-style config.

**Comment by SirCmpwn** [comment]
Alright, I'll retest and rereview this when I have some time (probably today or tomorrow). I'll try to get some more info about the swaygrab issue as well (fwiw the failure was when trying to capture video).

**Comment by thejan2009** [comment]
I'm not sure how Awesome does it, but you might have to look into container movements. The following was achieved by moving a container "up" into the bigger area and then moving it back "down".  When additional views are invoked, their width is different from every other container. I suppose they should be equally wide?
2017-01-08-190758_swaygrab

File: sway/sway.5.txt [hunk 19]
```diff
@@ -81,10 +84,28 @@ They are expected to be used with **bindsym** or at runtime through **swaymsg**(
 
 **layout** <mode>::
 	Sets the layout mode of the focused container. _mode_ can be one of _splith_,
-	_splitv_, _toggle split_, _stacking_ or _tabbed_.
+	_splitv_, _toggle split_, _stacking_, _tabbed_.
+
+**layout** auto <mode>::
+	Sets layout to one of the auto modes, i.e. one of _left_, right_, _top_,
+	or _bot_.
```
Comment (SirCmpwn) [inline sway/sway.5.txt]: Just spell out "bottom" here

File: sway/sway.5.txt [hunk 20]
```diff
@@ -360,7 +381,7 @@ The default colors are:
 	switch to workspace 2, then invoke the "workspace 2" command again, you
 	will be returned to workspace 1. Defaults to _no_.
 
-**workspace_layout** <default|stacking|tabbed>::
+**workspace_layout** <default|stacking|tabbed|auto_left|auto_right|auto_top|auto_bottom>::
```
Comment (SirCmpwn) [inline sway/sway.5.txt]: `workspace_layout auto left`, not `workspace_layout auto_left`

**Comment by Airblader** [comment]
(Random side comment: instead of making i3 jump through hoops for KDE integration KDE should just behave better to not require special casing. Specifications exist for good reasons. I'm not sure why people don't consider this and then blame i3 for it.)

File: sway/sway.5.txt (hunk 20 above)
Comment (willakat) [inline sway/sway.5.txt]: done

File: sway/sway.5.txt (hunk 19 above)
Comment (willakat) [inline sway/sway.5.txt]: done

**Comment by willakat** [comment]
Regarding **thejan2009**'s comment, the most recent batch of commits should address the issue. When inserting/removing a child in a container, the auto-layout will attempt to preserve sizes along the "major" axis. Along the "minor" axis, sizes are preserved for children that aren't promoted/demoted to a different group. In the latter case their size will match the spot vacated in the arrival group.

File: samples/awesome-like.config [hunk 21]
```diff
@@ -0,0 +1,63 @@
...
+#
+mode "resize" {
+    bindsym Left resize shrink width 20 px
+    bindsym Down resize grow height 20 px
+    bindsym Up resize shrink height 20 px
+    bindsym Right resize grow width 20 px
+}
+bindsym $mod+r mode "resize"
+
+new_window pixel 1
```
Comment (SirCmpwn) [inline samples/awesome-like.config]: Can you put this into a directory called `contrib` instead?

**Comment by SirCmpwn** [comment]
I would also make your sample awesome config more awesome like instead of trying to keep sway's default keybindings intact. Your config doesn't actually even include sway's default keybindings anyway, the defaults are via your config (rather than being hardcoded into sway). Fix that in a seperate pull request if you like.

**Comment by SirCmpwn** [comment]
Thanks for all the hard work!

**Comment by willakat** [comment]
You're welcome! Many thanks for yours on Sway and to the other contributors.

**Pull request merged**